src/back/msl/writer.rs

```diff
@@ -689,10 +689,58 @@ impl<W: Write> Writer<W> {
                 write!(self.out, "{}", name)?;
             }
             crate::Expression::Load { pointer } => {
-                // We don't do any dereferencing with `*` here as pointer arguments to functions
-                // are done by `&` references and not `*` pointers. These do not need to be
-                // dereferenced.
-                self.put_expression(pointer, context, is_scoped)?;
+                // Because packed vectors such as `packed_float3` cannot be directly multipied by
+                // matrices, we wrap them with `float3` on load.
+                let wrap_float_scalar_kind = match context.function.expressions[pointer] {
+                    crate::Expression::AccessIndex { base, index } => {
+                        let ty = match context.resolve_type(base) {
+                            &crate::TypeInner::Pointer { base, .. } => {
+                                &context.module.types[base].inner
+                            }
+                            ty => ty,
```

Comment:
I assume this path is not really expected here, but we just don't want to bother with any early exits. Maybe worth leaving a small comment?